### PR TITLE
Openshift userdata plugin

### LINKED ISSUE #3118: OS: Allow deployment on all supported providers
Currently its only possible to create Openshift clusters on Hetzner. We should extend this to allow creating them on all providers the machine-controller supports.

*Acceptance criteria*

* The openshift controller doesn't deploy a fork of the machine-controller but the upstream version
* Openshift clusters can be created on all providers that can be used for Kubernetes clusters

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3118

**Special notes for your reviewer**:

**Documentation**:

**Release note**:

```release-note
none
```

/hold

### CODE REVIEW

File: api/cmd/openshift-userdata/main.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package main
```
Comment (themue) [inline api/cmd/openshift-userdata/main.go]: Why a different path scheme for this package, not i.e. `api/cmd/userdata/openshift`?

Question: How will it be named as binary? Important for todays way of loading.

Comment (alvaroaleman) [reply api/cmd/openshift-userdata/main.go]: The main reason is that we have one maketarget for all binaries that names the binary like the directory the main.go is in and I dislike `openshift` as binary name.

This is not the name the binary has after being copied into the machine-controller container, see `pkg/controller/openshift/reddources/machinecontroller.go` for details

**Comment by alvaroaleman** [comment]
@themue This is ready now, PTAL

/hold cancel

File: api/Makefile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 export CGO_ENABLED?=0
-REPO=kubermatic/api
+REPO=quay.io/kubermatic/api
```
Comment (themue) [inline api/Makefile:2]: What's the motivation here?

File: api/hack/run-controller.sh [hunk 3]
```diff
@@ -3,6 +3,7 @@
 set -o errexit
 set -o nounset
 set -o pipefail
+set -x
```
Comment (themue) [inline api/hack/run-controller.sh:6]: Still needed?

File: api/pkg/controller/openshift/resources/machinecontroller.go [hunk 4]
```diff
@@ -1,52 +1,54 @@
...
-			if d.Spec.Template.Spec.Containers[idx].Name == containerName {
-				d.Spec.Template.Spec.Containers[idx].Image = image
-				containerWasFound = true
-				break
+		return name, func(in *appsv1.Deployment) (*appsv1.Deployment, error) {
+			d, err := creator(in)
+			if err != nil {
+				return nil, err
 			}
+			d.Spec.Template.Spec.Volumes = append(d.Spec.Template.Spec.Volumes, corev1.Volume{Name: "userdata-plugins",
```
Comment (themue) [inline api/pkg/controller/openshift/resources/machinecontroller.go:26]: That's quite interesting. Havn't yet thought of creating/initialising this way. Have to take a deeper look into the Go docs for the API here.

**Review by themue — COMMENTED** [review]
Only one question, rest LGTM.

/lgtm

File: api/Makefile (hunk 2 above)
Comment (alvaroaleman) [reply api/Makefile:2]: We don't use the dockerhub repo anymore

File: api/hack/run-controller.sh (hunk 3 above)
Comment (alvaroaleman) [reply api/hack/run-controller.sh:6]: Yes, because there is a long time of no feedback at all when something goes wrong during startup and this is not set

**Comment by alvaroaleman** [comment]
@themue Can you `/approve`? :)

**Comment by themue** [comment]
Sure, after these two helpful answers.

/approve

**Pull request merged**